File: C++/src/CellLoop.cc
#include"../include/CellLoop.h"

static void cc(double &aL, double &aD, double &aU, double &b
      , double rho, double cp, double kP, double kV
      , double dx, double dt, double u
      , short ccType, double *aCcValue, short c);


/**********************************************************************
 * Data de Ciacao:       18/04/2021                                   *
 * Data de Modificacao : 22/04/2021                                   *
 * -------------------------------------------------------------------*
 * montaSistema: loop nas celulas para montar os sistema de equacoes  *
 * -------------------------------------------------------------------*
 * Parametros de Entrada:                                             *
 * -------------------------------------------------------------------*
 * -------------------------------------------------------------------*
 * Parametros de Saida:                                               *
 * -------------------------------------------------------------------*
 * ------------------------------------------------------------------ *
 * Objetos utilizados:                                                *
 *------------------------------------------------------------------  *
 * DataStrucr:                                                        *
 *  |_ l[neq] - diagonal inferior  (alterado)                         *
 *  |_ d[neq] - diagonal principal (alterado)                         *
 *  |_ u[neq] - diagonal superio   (alterado)                         *
 *  |_ b[neq] - vetor independente (alterado)                         *
 * Mesh:                                                              *
 * |_ cells:                                                          *
 *    |_ u - valores da solucao por celula                            *
 * |  |_ prop:                                                        *
 * |     |_ rho - massa especifica                                    *
 * |     |_ cp  - calor especifico                                    *
 * |     |_ k   - coeficiente de difusao                              *
 * |_ cccci:                                                          *
 *       |_ cceType  - tipo de condicao de contorno esquerda          *
 *       |_ cceValue - valor de condicao de contorno esquerda         *
 *       |_ ccdType  - tipo de condicao de contorno direita           *
 *       |_ ccdValue - valor de condicao de contorno esquerda         *
 * intTemp                                                           *
 * |_ dt - delta t                                                    *
 * ------------------------------------------------------------------ *
 * OBS:                                                               *
 * ------------------------------------------------------------------ *
 *********************************************************************/
void CellHeatLoop::montaSistema(void){
  // ...
  double aP0, kf, aE, aW;
  // ..
  double *rho = this->mesh->getCells().getProp().get_rho();
  double *cp = this->mesh->getCells().getProp().get_cp();
  double *k = this->mesh->getCells().getProp().get_k();
  // ...
  double dt = this->intTemp->get_dt();
  double dx = this->mesh->getCells().get_dx();
  // ...
  int cceType = this->mesh->getCcci().get_cceType(),
      ccdType = this->mesh->getCcci().get_ccdType();
  double *cceValue = this->mesh->getCcci().get_cceValue(),
         *ccdValue = this->mesh->getCcci().get_ccdValue();
  double *u = this->mesh->getCells().getPu();
  // ... sistema de equacoes
  double *aU = this->solver->get_dataStruct()->get_a3(),
         *aD = this->solver->get_dataStruct()->get_a2(),
         *aL = this->solver->get_dataStruct()->get_a1(),
          *b = this->solver->get_dataStruct()->get_b();
  // ...
  int nCells = this->mesh->get_nCells(), n;
  // ..........................................................................

  // ... Lado esquerdo
  cc(aL[0]  , aD[0], aU[0], b[0],
     rho[0] , cp[0], k[0] , k[1],
     dx     , dt   , u[0],  
     cceType, cceValue, cce);
  // ............................................................................


  // ... Lado direito
  n = nCells - 1;
  cc(aL[n] , aD[n]   , aU[n], b[n],
    rho[n] , cp[n]   ,  k[n], k[n-1],
    dx     , dt      , u[n],
    ccdType, ccdValue, ccd);
  // ............................................................................
 
  // ... loop nas celulas do interios
  for (int i = 1; i < nCells - 1; i++) {
    aP0 = rho[i] * cp[i] * dx / dt;
    // ... w
    kf = (k[i - 1] + k[i])*0.5e0;
    aW = kf / dx;
    // ...
    kf = (k[i] + k[i + 1])*0.5e0;
    aE = kf / dx;
    // ...
    aL[i] = -aW;
    aD[i] = aP0 + aW + aE;
    aU[i] = -aE;
    // ...
    b[i] = aP0 * u[i];
  }
  // ..........................................................................
}
//*****************************************************************************

/********************************************************************************
 * Data de Ciacao:       22/04/2021                                             *
 * Data de Modificacao : 00/00/0000                                             *
 * ---------------------------------------------------------------------------- *
 * cc: aplicao da condicoes de contorno                                         *
 * ---------------------------------------------------------------------------- *
 * Parametros de entrada:                                                       *
 * ---------------------------------------------------------------------------- *
 * aL      - termo diagonal inverior                                            *
 * aD      - termo diagonal principal                                           *
 * aU      - termo diagonal superio                                             *
 * b       - termo vetor de forcas                                              *
 * rho     - massa especifica                                                   *
 * cp      - calor especifico                                                   *
 * kP      - coeficiente de difusao da celula central                           *
 * kV      - coeficiente de difusao da celula vizina                            *
 * h       - coeficiente de calor convectivo                                    *
 * dx      - tamanho da celula                                                  *
 * dt      - passo de tempo                                                     *
 * u       - valor da solução do passo de tempo anterior                        *
 * ccType  - tipo da condicao de contorno                                       *
 * ccValue - parametros da condicao de contorno                                 *
 * ---------------------------------------------------------------------------- *
 * Parametros de saida:                                                         *
 * ---------------------------------------------------------------------------- *
 * aL      - termo diagonal inverior                                            *
 * aD      - termo diagonal principal                                           *
 * aU      - termo diagonal superio                                             *
 * b       - termo vetor de forcas                                              *
 * ---------------------------------------------------------------------------- *
 * Objetos utilizados:                                                          *
 *----------------------------------------------------------------------------- *
 * ---------------------------------------------------------------------------- *
 * OBS:                                                                         *
 * ---------------------------------------------------------------------------- *
 ********************************************************************************/
static void cc(double &aL, double &aD, double &aU, double &b
             , double rho, double cp, double kP, double kV
             , double dx, double dt, double u
             , short ccType, double *aCcValue, short c) {

  double aP0, sU, sP, kf, aWorE;
  double ccValue = aCcValue[0];

  aP0 = rho * cp * dx / dt;
  kf = (kP + kV)*0.5e0;
  aWorE = kf / dx;
  // ... temperatura prescrita
  if (ccType == typeCc::temp) {
    sU = (2.0e0*kP / dx)*ccValue;
    sP = -2.0e0*kP / dx;
  }
  // ... fluxo prescrito
  else if (ccType == typeCc::flux) {
    sU    = -ccValue;
    sP    = 0.0;
  }
  // ... lei de resfriamento de newton
  else if (ccType == typeCc::hConv) {
    double h = aCcValue[1];
    double tmp = 1.e0 + (h*2.e0*dx)/kP;
    tmp = h/tmp;
    sU = tmp*ccValue;
    sP = -tmp;
  }
  // ............................................................................

  // ... W
  aL = -((c == cce) ? -0.0 : aWorE);
  // ... P
  aD = aP0 + aWorE - sP;
  // ... E
  aU = -((c == ccd) ? -0.0 : aWorE);
  // c ...
  b = sU + aP0 * u;
  // ............................................................................
}
/******************************************************************************/

File: Fortran/src/CellLoop.f
c *********************************************************************
c * DATA DE CRIACAO  : 11/09/2018                                     *
c * DATA DE MODIFICAO: 00/00/0000                                     *
c * ----------------------------------------------------------------- *
c * MONTASISTEMA : monta o sistema AX=B                               *
c * ----------------------------------------------------------------- *
c * Parametros de entrada:                                            *
c * ----------------------------------------------------------------- *
c * a      - nao definido                                             *
c * b      - nao definido                                             *
c * temp   - temperatura do passo anterior                            *
c * sQ     - fonte de calor                                           *
c * k      - coeficiente de conducao termica                          *
c * ro     - massa especidica                                         *
c * cp     - calor espefifico                                         *
c * dt     - delta t                                                  *
c * cc     - tipo de condicao de contorno                             * 
c         1 - Temperatuta                                             *
c         2 - fluxo                                                   *
c         3 - resfriamento de newton                                  *
c * ccv    - valor numerico da condicao de contorno                   *
c * dx     - delta x                                                  *
c * nCells - numero de celulas (elementos)                            * 
c * ----------------------------------------------------------------- *
c * Parametros de saida:                                              *
c * ----------------------------------------------------------------- *
c * a    - matriz de coefientes                                       *
c * b    - vetor b                                                    *
c * ----------------------------------------------------------------- *
c * OBS:                                                              *
c * ----------------------------------------------------------------- *
c *********************************************************************
      subroutine  montaSistema(a,b,temp,sQ,k,ro,cp,dt,cc,ccv,h,dx
     1                        ,nCells)
      implicit none
      integer cc(2),nCells,i
      real(8) a(nCells,*),b(*),k(*),ro(*),cp(*),temp(*),sQ(*)
      real(8) dt,ccv(2),dx,sP,sU,kf,tmp,h(2)
      real(8) aE,aW,aP0
c ... temperatura prescrita
      if(cc(1) .eq. 1) then
        aP0    = ro(1)*cp(1)*dx/dt
        sU     = (2.0d0*k(1)/dx)*ccv(1)
        sP     = -2.0d0*k(1)/dx
        kf     = (k(1)+k(2))*0.5d0
        aE     = kf/dx
c ... W
        a(1,1) = 0.0
c ... P
        a(1,2) = aP0 + aE - sP
c ... E
        a(1,3) = -aE
c ...
        b(1)   = sU + aP0*temp(1)
c ... fluxo prescrito
      else if(cc(1) .eq. 2) then
        aP0    = ro(1)*cp(1)*dx/dt
        sU     = -ccv(1)
        sP     = 0.0   
        kf     = (k(1)+k(2))*0.5d0
        aE     = kf/dx
c ... W
        a(1,1) = 0.0
c ... P
        a(1,2) = aP0 + aE - sP
c ... E
        a(1,3) = -aE
c ...
        b(1)   = sU + aP0*temp(1)
c ... lei de resfriamento de newton
      else if(cc(1) .eq. 3) then
        tmp    = 1.0d0 + (h(1)*2.0*dx)/k(1)
        tmp    = h(1)/tmp
        aP0    = ro(1)*cp(1)*dx/dt
        sU     = tmp*ccv(1)
        sP     = -tmp   
        kf     = (k(1)+k(2))*0.5d0
        aE     = kf/dx
c ... W
        a(1,1) = 0.0
c ... P
        a(1,2) = aP0 + aE - sP
c ... E
        a(1,3) = -aE
c ...
        b(1)   = sU + aP0*temp(1)

      endif
c .....................................................................

c ... temperatura prescrita
      if(cc(2) .eq. 1) then
        aP0    = ro(nCells)*cp(nCells)*dx/dt
        sU     = (2.0d0*k(nCells)/dx)*ccv(2)
        sP     = -2.0d0*k(nCells)/dx
        kf     = (k(nCells-1)+k(nCells))*0.5d0
        aW     = kf/dx
c ... W
        a(nCells,1) = -aW
c ... P
        a(nCells,2) = aP0 + aW - sP
c ... E
        a(nCells,3) = 0.0
c ...
        b(nCells)   = sU + aP0*temp(nCells)
c ... fluxo prescrito
      else if(cc(2) .eq. 2) then
        aP0    = ro(nCells)*cp(nCells)*dx/dt
        sU     = -ccv(2)
        sP     = 0.0
        kf     = (k(nCells-1)+k(nCells))*0.5d0
        aW     = kf/dx
c ... W
        a(nCells,1) = -aW
c ... P
        a(nCells,2) = aP0 + aW - sP
c ... E
        a(nCells,3) = 0.0
c ...
        b(nCells)   = sU + aP0*temp(nCells)
c ... lei de resfriamento de newton
      else if(cc(2) .eq. 3) then
        tmp    = 1.0d0 + (h(2)*2.0*dx)/k(nCells)
        tmp    = h(2)/tmp
        aP0    = ro(nCells)*cp(nCells)*dx/dt
        sU     = tmp*ccv(2)
        sP     = -tmp   
        kf     = (k(nCells-1)+k(nCells))*0.5d0
        aW     = kf/dx
c ... W
        a(nCells,1) = -aW
c ... P
        a(nCells,2) = aP0 + aW - sP
c ... E
        a(nCells,3) = 0.0
c ...
        b(nCells)   = sU + aP0*temp(nCells)
      endif
c .....................................................................

c ... loop nas celulas do interios
      do i = 2, nCells -1
        aP0    = ro(i)*cp(i)*dx/dt
c ... w 
        kf = (k(i-1) + k(i))*0.5d0
        aW = kf/dx  
c ...
        kf = (k(i) + k(i+1))*0.5d0
        aE = kf/dx  
c ... 
        a(i,1) = -aW 
        a(i,2) = aP0 + aW + aE 
        a(i,3) = -aE
c ... 
        b(i)   = aP0*temp(i)
      enddo
c .....................................................................
      return
      end
c *********************************************************************
